### ISSUE TITLE
BUG: interp1d returns NaNs when extrapolating with "previous" and "next" methods

### ISSUE THREAD

**Issue by lschr** [scipy/scipy#16997]
### Describe your issue.

Since version 1.9.0, NaNs are returned when interpolating values
- larger than the greatest x if `kind="next"`
- less than the smallest x if `kind="previous"`

Previously, the y value corresponding to the largest x and the y value corresponding to the smallest x, respectively, were returned.

### Reproducing Code Example

```python
scipy.interpolate.interp1d([0, 1, 2, 3], [1, 2, 3, 4], "next", fill_value="extrapolate")([-1, 1.5, 5])
# returns [ 1.,  3., nan], was [1., 3., 4.] prior to version 1.9

scipy.interpolate.interp1d([0, 1, 2, 3], [1, 2, 3, 4], "previous", fill_value="extrapolate")([-1, 1.5, 5])
# returns [nan,  2.,  4.], was [1., 2., 4.] prior to version 1.9
```

### Error message

```shell
No error message.
```

### SciPy/NumPy/Python version information

1.9.1 1.22.4 sys.version_info(major=3, minor=9, micro=13, releaselevel='final', serial=0)

**Comment by tylerjereddy** [comment]
The patch below the fold seems to restore the behavior you want, but it would contradict the testing changes added in gh-16825, so we need to ping @ev-br and @AtsushiSakai for clarification.

<details>

```diff
--- a/scipy/interpolate/_interpolate.py
+++ b/scipy/interpolate/_interpolate.py
@@ -528,7 +528,7 @@ class interp1d(_Interpolator1D):
                 if _do_extrapolate(fill_value):
                     self._check_and_update_bounds_error_for_extrapolation()
                     # assume y is sorted by x ascending order here.
-                    fill_value = (np.nan, np.take(self.y, -1, axis))
+                    fill_value = (np.take(self.y, 0, axis), np.take(self.y, -1, axis))
             elif kind == 'next':
                 self._side = 'right'
                 self._ind = 1
@@ -538,7 +538,7 @@ class interp1d(_Interpolator1D):
                 if _do_extrapolate(fill_value):
                     self._check_and_update_bounds_error_for_extrapolation()
                     # assume y is sorted by x ascending order here.
-                    fill_value = (np.take(self.y, 0, axis), np.nan)
+                    fill_value = (np.take(self.y, 0, axis), np.take(self.y, -1, axis))
             else:
                 # Check if we can delegate to numpy.interp (2x-10x faster).
                 np_types = (np.float_, np.int_)
diff --git a/scipy/interpolate/tests/test_interpolate.py b/scipy/interpolate/tests/test_interpolate.py
index 7bcfdadd8..dbb66935b 100644
--- a/scipy/interpolate/tests/test_interpolate.py
+++ b/scipy/interpolate/tests/test_interpolate.py
@@ -437,6 +437,12 @@ class TestInterp1D:
                          [10, 11],
                          [-30, -30]])
 
+
+    def test_previous_gh_16997(self):
+        actual = interp1d([0, 1, 2, 3], [1, 2, 3, 4], "previous", fill_value="extrapolate")([-1, 1.5, 5])
+        assert_allclose(actual, [1., 2., 4.])
+
+
     def test_next(self):
         # Check the actual implementation of next interpolation.
         interp10 = interp1d(self.x10, self.y10, kind='next')
@@ -506,6 +512,13 @@ class TestInterp1D:
                          [10, 11],
                          [np.nan, np.nan]])
 
+
+    def test_next_gh_16997(self):
+        actual = interp1d([0, 1, 2, 3], [1, 2, 3, 4],
+                          "next", fill_value="extrapolate")([-1, 1.5, 5])
+        assert_allclose(actual, [1., 3., 4.])
+
+
     def test_zero(self):
         # Check the actual implementation of zero-order spline interpolation.
         interp10 = interp1d(self.x10, self.y10, kind='zero')
```

</details>

I pushed the draft changes up to my fork on branch `treddy_issue_16997` if they're desired for additional development.

**Comment by AtsushiSakai** [comment]
I think the new behavior is expected from v1.9 as you said, and 
> Previously, the y value corresponding to the largest x and the y value corresponding to the smallest x, respectively, were returned.

is a defect as reported in #9591. We fixed the behavior in v1.9.
Because, 
> larger than the greatest x if kind="next"

This means there is no next `x`, so `nan` should be returned.

> less than the smallest x if kind="previous"

This means there is no previous `x`, so `nan` should be returnd.
Any comments? @lschr

**Comment by lschr** [comment]
Sorry, I completely missed #9591. Since the new behavior seems intended and it is easy to get back the old behavior via `bounds_error=False, fill_value=(y[0], y[-1])`, I guess this can be closed.
